src/http/client.cc

```diff
@@ -323,11 +324,24 @@ struct transport_error : public std::exception {
 
 future<> client::make_request(request req, reply_handler handle, reply::status_type expected) {
     return do_with(std::move(req), std::move(handle), [this, expected] (request& req, reply_handler& handle) mutable {
-        return do_make_request(req, handle, expected);
+        return do_make_request(req, handle, expected, force_new_connection::no).handle_exception_type([this, &req, &handle, expected] (const transport_error& ex) {
+            if (ex.requests_submitted > 1 /* cached connection */) {
+                // This forces one more connection without checking the limit, but
+                // since the old connection is closing it compensates the increase.
+                //
+                // However, in case new make_request() manages to creep in between
+                // it will temporarily create one extra connection, but it will not
+                // be held in pool anyway. Still, this is unlikely, _nr_connections
+                // decrease will happen in a waiting task and this place is chained
+                // directly after the put_connection()
+                return do_make_request(req, handle, expected, force_new_connection::yes);
```

Comment:
> But why does it need to be a new connection? Why can't we reuse an existing connection already lying in the pool and it must be a new connection? Why will this new connection "not be held in the pool anyway" - we'll only use it once?

That's _exactly_ the question we're trying to resolve. There are currently two implementation of "retry failed http requests". This PR and the #1847 one. This PR retries request over fresh new connection. This approach is good, because it doesn't have any magic parameters from the caller (well, maybe a boolean flag whether to do it at all or not) -- it _just_ retries the request once over new connection and that's it. THe #1847 implements request retrying in the "backoff-retry" manner. I.e. if a request fails it waits for a while, then picks up another connection (and in that case this another connection can be from pool or can be a new one, depending on the pool state) and tries request again. And does this several times. That approach needs configuration from the caller -- the number of retrying attempts and the pause duration between retries.

> Also, I don't understand the discussion about concurrent requests and limit. If a request is made, and there is no open connection, we have to create a new connection - we can't "limit" this number. 

We don't

> We can limit the number of unused connections we cache, but how is it relevant here?

It just says that when the do_make_request() failed the connection that served that request is still alive at this place and it will only get closed few continuations after that. By the time this connection gets destroyed, request retrying would likely create a _new_ connection _without_ checking the limit thus breaking this limit for a short period of time. It's not critical, just a clarification comment.